### PR TITLE
New Command: `make:document`

### PR DESCRIPTION
**Pull request by constantable**
New Command: `make:document` to interactively generate document classes for `doctrine/mongodb-odm-bundle`

Related issues #320 , #364

### CODE REVIEW

**Comment by constantable** [comment]
Need help with the MongoDB and ext-mongodb PHP extention setup for AppVeyor.

**Comment by constantable** [comment]
> Need help with the MongoDB and ext-mongodb PHP extention setup for AppVeyor.

Fixed AppVeyor

**Comment by OskarStark** [comment]
cc @alcaeus

File: src/Doctrine/DoctrineODMHelper.php [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+    }
+
+    private function isInstanceOf($object, string $class): bool
+    {
+        if (!\is_object($object)) {
+            return false;
+        }
+
+        return $object instanceof $class;
+    }
```
Comment (alcaeus) [inline src/Doctrine/DoctrineODMHelper.php:296]: `$value instanceof $class` works even when `$value` is not an object. Not sure why you'd need this helper.

File: src/Doctrine/DoctrineODMHelper.php [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+    public function getPotentialCollectionName(string $className): string
+    {
+        $documentManager = $this->getRegistry()->getManager();
+
+        if (!$documentManager instanceof DocumentManager) {
+            throw new \RuntimeException('ObjectManager is not an DocumentManager.');
```
Comment (alcaeus) [inline src/Doctrine/DoctrineODMHelper.php:303]:
```suggestion
            throw new \RuntimeException('ObjectManager is not a DocumentManager.');
```

File: src/Doctrine/DoctrineODMHelper.php [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+
+        return $className;
+    }
+
+    public function isKeyword(string $name): bool
+    {
+        /** @var Connection $connection */
+        $connection = $this->getRegistry()->getConnection();
+
+        return $connection->getDatabasePlatform()->getReservedKeywordsList()->isKeyword($name);
```
Comment (alcaeus) [inline src/Doctrine/DoctrineODMHelper.php:323]: There are no such things as reserved keywords in MongoDB.

File: src/Doctrine/DocumentRegenerator.php [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+            $manipulator = $this->createClassManipulator($classPath);
+            $operations[$classPath] = $manipulator;
+
+            $embeddedClasses = [];
+
+            foreach ($classMetadata->getEmbeddedFieldsMappings() as $fieldName => $mapping) {
+                if (str_contains($fieldName, '.')) {
+                    continue;
+                }
```
Comment (alcaeus) [inline src/Doctrine/DocumentRegenerator.php:77]: While not recommended, field names in MongoDB can contain dots. Why is this excluded here?

File: src/Doctrine/DocumentRegenerator.php [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                        $relation = (new ReferenceMany(
+                            propertyName: $mapping['fieldName'],
+                            targetDocument: $mapping['targetDocument'],
+                            targetPropertyName: $mapping['mappedBy'],
+                            orphanRemoval: $mapping['orphanRemoval'],
+                        ));
+
+                        $manipulator->addOneToManyReference($relation);
+
+                        break;
```
Comment (alcaeus) [inline src/Doctrine/DocumentRegenerator.php:142]: Associations in ODM come in two variants: references and embedded documents. You assume that any association is a reference, which is incorrect. You need to differentiate on `$mapping['embedded']` to distinguish between references and embedded documents.

**Review by alcaeus — CHANGES_REQUESTED** [review]
I've left some code feedback below.

What I'm completely missing in this PR is support for embedded documents. For one, this should support creating embedded document classes (which use a different document attribute, have no repository, and can be final), as well as creating embedded relationships in the maker itself. Currently the maker assumes all relationships to be references, which isn't true for ODM. In fact, embedded documents are a core concept in MongoDB, and as such I believe they need to be included in the builder for it to be considered.

**Comment by constantable** [comment]
@alcaeus Thank you for code review and suggestions!

Implemented support for embedded documents. 
An EmbeddedDocument could be created form scratch if `--embedded` option specified. 

When embeding a class that has the `ODM\Document` PHP attribute maker will ask should it be transformed to `ODM\EmbeddedDocument`. Please leave a word what you think about this approach.

File: src/Doctrine/DoctrineODMHelper.php (hunk 1 above)
Comment (constantable) [reply src/Doctrine/DoctrineODMHelper.php]: DoctrineODMHelper is mostly based on DoctrineHelper as you noted. I have removed this method as minimum PHP version is 8.0 and `instanceof` acts exactly the same as the isInstanceOf method.
Am I supposed to make corresponding changes in DoctrineHelper in this PR?

**Comment by alcaeus** [comment]
> Implemented support for embedded documents. An EmbeddedDocument could be created form scratch if `--embedded` option specified.

Sounds good!

> When embeding a class that has the `ODM\Document` PHP attribute maker will ask should it be transformed to `ODM\EmbeddedDocument`. Please leave a word what you think about this approach.

The problem is that anything mapped as document can't be embedded, and anything mapped as embedded document can't be stored as root document in a collection. IMO the smartest choice would be to print an error message and suggest an alternative (e.g. a mapped superclass containing common properties, extended by two separate classes that then add the `Document` and `EmbeddedDocument` mappings). Unfortunately, changing a `Document` to an `EmbeddedDocument` will break existing functionality if the document is already used in a non-embedded context somewhere. I believe it's fine to explain this problem to the user and letting them solve it themselves before allowing to embed the class.

**Comment by constantable** [comment]
> 

Made changes to print the error message.
image

**Comment by weaverryan** [comment]
Hi there!

I think this is a very cool thing. I have avoided reviewing it because... to be fully honest, I don't want to have to maintain all of the code to make this work. `make:entity` requires a lot of code and complexity, but I'm ok with it because I use it. But I won't use this command (nothing bad about the ODM! It's just not in my stack).

I would prefer to have this as a 3rd party, independent bundle and to add any new APIs to MakerBundle to make that possible. I would even be cool with adding a fake` make:document` to MakerBundle that tells you what `composer require` command to run to get the bundle that *actually* adds it.

Let me know what you think - and sorry for not talking with you about this sooner - it's always very busy.

Cheers!

**Comment by jrushlow** [comment]
Howdy @constantable - this looks awesome! But I have to agree with @weaverryan on the maintenance burden this would add to maker-bundle. If you need help on getting this going as a 3rd party bundle, please don't hesitate to create an issue, we'll be glad to help.

We love new ideas and the PR's that are created from those ideas, so keep them coming! Thanks...

**Comment by TracKer** [comment]
@constantable It would be really cool if you could organize your code into a separate maker-bundle package

File: src/Maker/MakeDocument.php [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+use Symfony\Bundle\MakerBundle\Doctrine\ODMDependencyBuilder;
+use Symfony\Bundle\MakerBundle\Exception\RuntimeCommandException;
+use Symfony\Bundle\MakerBundle\FileManager;
+use Symfony\Bundle\MakerBundle\Generator;
+use Symfony\Bundle\MakerBundle\InputAwareMakerInterface;
+use Symfony\Bundle\MakerBundle\InputConfiguration;
+use Symfony\Bundle\MakerBundle\Str;
+use Symfony\Bundle\MakerBundle\Util\ClassDetails;
+use Symfony\Bundle\MakerBundle\Util\ClassSourceManipulator;
+use Symfony\Bundle\MakerBundle\Util\PhpCompatUtil;
```
Comment (GromNaN) [inline src/Maker/MakeDocument.php:33]: This classes are marked as internal, in order to create a 3rd party package, we would have to duplicate them.

File: src/Doctrine/ODMDependencyBuilder.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+/*
+ * This file is part of the Symfony MakerBundle package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\MakerBundle\Doctrine;
```
Comment (GromNaN) [inline src/Doctrine/ODMDependencyBuilder.php:12]: In order to ease maintenance, all classes related to MongoDB ODM should be in a specific namespace: `Symfony\Bundle\MakerBundle\Doctrine\MongoDB`

**Review by GromNaN — COMMENTED** [review]
Due to the rejection of this PR, we will create a new bundle for Doctrine MongoDB ODM support with MakerBundle.
https://github.com/doctrine/mongodb-maker-bundle

**Comment by constantable** [comment]
@GromNaN you can use https://github.com/constantable/odm-document-maker if you find it useful.

**Pull request closed without merging**